### PR TITLE
2G - part 1 of 2

### PR DESCRIPTION
**Pull request by matthayter**
Reopening with all comments addressed from #7

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -40,7 +40,7 @@ The metrics library must be included before being used.
 mc = require("com.arpnetworking.metrics.metrics-client")
 ```
 
-This will write the metrics to file named ```tsd-query.log``` with rotation size of 32MB and retention of the last 10 logs
+This will write the metrics to file named ```query.log``` with rotation size of 32MB and retention of the last 10 logs
```
Comment (vjkoskela) [inline README.md]: Does this still match the defaults?

File: src/sinks/tsd-query-log-sink.ts [hunk 2]
```diff
@@ -72,29 +74,35 @@ export class TsdQueryLogSink implements tsdDef.Sink {
     private static stringify(metricsEvent:tsdDef.MetricsEvent):string {
         var transformedMetricsEvent = TsdQueryLogSink.transformMetricsEvent(metricsEvent);
```
Comment (vjkoskela) [inline src/sinks/tsd-query-log-sink.ts:76]: So in 2F we supported both a standard json format and a steno json format. I know you are skipping 2f, but this looks like the non-steno sink in Java and it implements steno.

File: src/tsd-metrics-factory.ts [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+        if (clusterName == null) {
+            failures.push("clusterName must be provided");
+        }
+        if (serviceName == null) {
+            failures.push("serviceName must be provided");
+        }
+        if (hostResolver == null) {
+            if (options.hostName != null) {
+                hostResolver = new hostResolvers.DefaultHostResolver(options.hostName)
```
Comment (vjkoskela) [inline src/tsd-metrics-factory.ts]: Doesn't the default host resolver get the hostname from the operating system?

-> Consider a different "static" host provider for representing a user specified host name.

File: src/tsd-default-options.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    /* istanbul ignore next */
+    constructor() {
+    }
+
+    /**
+     * Sets the maximums size of log in bytes before rolling a new file.
+     * Default: 100 MB
+     * @type {number}
+     */
+    public static LOG_MAX_SIZE:number = 100 * 1024 * 1024;
```
Comment (vjkoskela) [inline src/tsd-default-options.ts:31]: Please add a TODO somewhere appropriate referencing a (new) issue that the NodeJS client needs to implement time-based file rotation.

File: src/tsd-metrics-factory.ts [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+import Options = options.Options;
+
+/**
+ * Factory pattern for creating and configuring instances of Metrics.
+ *
+ * @class
+ * @author Matthew Hayter (mhayter at groupon dot com)
+ */
+export class TsdMetricsFactory implements tsdDef.MetricsFactory {
+    private static defaultHostResolver = new hostResolvers.DefaultHostResolver(os.hostname());
```
Comment (vjkoskela) [inline src/tsd-metrics-factory.ts]: You should have two host name providers:

1) Accepts a static hostname and serves that back.
2) Returns the hostname from the os. 

Calling the OS from here defeats the point of the encapsulation.

If you want to implement more depth here, we have two additional host providers in Java:

1) Caching; caches the os hostname for a period of time but refreshes inline (e.g. when user calls getHostname).
2) Background refresh; caches the hostname and uses a thread (in JS a timer and handler) to refresh the cached name periodically without blocking user requests.

File: src/tsd-metrics-factory.ts [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+    private _hostResolver:tsdDef.HostResolver;
+    private _sinks:tsdDef.Sink[];
+    private _logger:log4js.Logger;
+
+    /**
+     * Build an instance of <code>TsdMetricsFactory</code> by providing the full configuration.
+     *
+     * @returns {TsdMetricsFactory}
+     */
+    public static buildInstance(options: {serviceName:string; clusterName:string; hostName?:string; hostResolver?:tsdDef.HostResolver; sinks?:tsdDef.Sink[]; logger?:log4js.Logger}):TsdMetricsFactory {
```
Comment (vjkoskela) [inline src/tsd-metrics-factory.ts:53]: Why would you accept the host name and host resolver in the method? Is the idea that only one is provided (e.g. arguments by name)?

**Review by vjkoskela — CHANGES_REQUESTED** [review]
Okay, just some minor feedback and responses to the closed PR (#7). I think after this we can merge this, but there's still a lot left to do before this can be released. This is neither 2F nor 2G.

File: README.md (hunk 1 above)
Comment (matthayter) [reply README.md]: Fixing...

File: src/sinks/tsd-query-log-sink.ts (hunk 2 above)
Comment (matthayter) [reply src/sinks/tsd-query-log-sink.ts:76]: Ah you're saying this should really be named `TsdQueryLogStenoSink` or similar?

As an aside, I'm planning to drop steno with the second part of these changes with 2G.

File: src/tsd-metrics-factory.ts (hunk 6 above)
Comment (matthayter) [reply src/tsd-metrics-factory.ts:53]: Correct: this method takes one argument called `options` which is an object with the specified interface. The interface specifies that both `hostName` and `hostResolver` are optional. If both are provided, `hostResolver` takes priority.

File: src/tsd-metrics-factory.ts (hunk 5 above)
Comment (matthayter) [reply src/tsd-metrics-factory.ts]: Oof, okay, mis-communication. Hopefully its what you're looking for now.

File: src/tsd-metrics-factory.ts (hunk 3 above)
Comment (matthayter) [reply src/tsd-metrics-factory.ts]: Oof, okay, mis-communication. Hopefully its what you're looking for now.

File: src/tsd-default-options.ts (hunk 4 above)
Comment (matthayter) [reply src/tsd-default-options.ts:31]: Done

**Comment by vjkoskela** [comment]
Looks good Matt, can't wait to see part 2.

**Pull request merged**